File: yahtzee.rb
#!/usr/bin/env ruby
# yahtzee.rb
# encoding: UTF-8

require 'json'
require 'fileutils'

COLORS = {
  reset: "\e[0m",
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  magenta: "\e[95m",
  cyan: "\e[96m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

CATEGORIES = [
  "единицы", "двойки", "тройки", "четвёрки", "пятёрки", "шестёрки",
  "три одинаковых", "четыре одинаковых", "фулл-хаус",
  "малый стрит", "большой стрит", "шанс", "яхтзи"
]
DICE_EMOJI = ['⚀', '⚁', '⚂', '⚃', '⚄', '⚅']
DICE_COLORS = [:red, :green, :yellow, :blue, :magenta, :cyan]

class Yahtzee
  attr_reader :mode, :players, :current, :dice, :rolls_left, :keep, :game_over, :stats_file

  def initialize(mode, names)
    @mode = mode
    @players = (names || (mode == 'ai' ? ['Игрок 1', 'Компьютер'] : ['Игрок 1', 'Игрок 2'])).map do |n|
      { name: n, score: 0, categories: {} }
    end
    @current = 0
    @dice = [0,0,0,0,0]
    @rolls_left = 3
    @keep = [false,false,false,false,false]
    @game_over = false
    @stats_file = File.join(Dir.home, '.yahtzee_stats.json')
    load_stats
  end

  def load_stats
    if File.exist?(@stats_file)
      @stats = JSON.parse(File.read(@stats_file))
    else
      @stats = { 'games' => 0, 'wins' => {} }
    end
  end

  def save_stats
    File.write(@stats_file, JSON.pretty_generate(@stats))
  end

  def display_stats
    if @stats['games'] == 0
      puts colorize("Статистика пуста.", :yellow)
      return
    end
    puts colorize("📊 Статистика:", :bold)
    puts "  Всего игр: #{@stats['games']}"
    @stats['wins'].each { |name, wins| puts "  #{name}: #{wins} побед" }
  end

  def roll_dice
    5.times { |i| @dice[i] = rand(1..6) unless @keep[i] }
    @rolls_left -= 1
  end

  def dice_display
    @dice.each_with_index.map { |d, i| colorize(DICE_EMOJI[d-1] + d.to_s, DICE_COLORS[d-1]) }.join(' ')
  end

  def show_dice
    puts "Кости: #{dice_display}"
  end

  def show_categories(player_idx)
    used = @players[player_idx][:categories]
    puts "\nДоступные категории:"
    num = 1
    CATEGORIES.each do |cat|
      puts "  #{num}. #{cat}" unless used[cat]
      num += 1
    end
  end

  def score_category(dice, cat)
    counts = Array.new(7, 0)
    total = 0
    dice.each { |d| counts[d] += 1; total += d }
    case cat
    when 'единицы' then counts[1] * 1
    when 'двойки' then counts[2] * 2
    when 'тройки' then counts[3] * 3
    when 'четвёрки' then counts[4] * 4
    when 'пятёрки' then counts[5] * 5
    when 'шестёрки' then counts[6] * 6
    when 'три одинаковых'
      return total if counts.any? { |c| c >= 3 }
      0
    when 'четыре одинаковых'
      return total if counts.any? { |c| c >= 4 }
      0
    when 'фулл-хаус'
      return 25 if counts.any? { |c| c == 3 } && counts.any? { |c| c == 2 }
      return 25 if counts.any? { |c| c == 5 }
      0
    when 'малый стрит'
      (1..3).each do |start|
        ok = true
        4.times { |i| ok = false if counts[start+i] == 0 }
        return 30 if ok
      end
      0
    when 'большой стрит'
      ok1 = true; ok2 = true
      (1..5).each { |i| ok1 = false if counts[i] == 0 }
      (2..6).each { |i| ok2 = false if counts[i] == 0 }
      return 40 if ok1 || ok2
      0
    when 'шанс' then total
    when 'яхтзи'
      return 50 if counts.any? { |c| c == 5 }
      0
    else 0
    end
  end

  def best_category(dice, player_idx)
    best = -1
    best_cat = nil
    used = @players[player_idx][:categories]
    CATEGORIES.each do |cat|
      next if used[cat]
      score = score_category(dice, cat)
      if score > best
        best = score
        best_cat = cat
      end
    end
    [best_cat, best]
  end

  def take_turn(player_idx)
    p = @players[player_idx]
    puts "\n#{colorize("Ход: " + p[:name], :bold)}"
    @rolls_left = 3
    @keep = [false,false,false,false,false]
    roll_dice
    show_dice

    while @rolls_left > 0
      if @mode == 'ai' && player_idx == 1
        counts = Array.new(7, 0)
        @dice.each { |d| counts[d] += 1 }
        max_cnt = counts.max
        target = counts.index(max_cnt)
        if max_cnt >= 2
          @keep = @dice.map { |d| d == target }
        else
          @keep = [false,false,false,false,false]
        end
        roll_dice
        show_dice
        break
      else
        print "Введите номера кубиков для удержания (через пробел, 0 - оставить все): "
        line = gets.chomp.strip
        exit if line == 'q'
        if line == '0'
          @keep = [true,true,true,true,true]
        else
          @keep = [false,false,false,false,false]
          line.split.each do |token|
            idx = token.to_i - 1
            @keep[idx] = true if idx >= 0 && idx < 5
          end
        end
        if @rolls_left > 1
          puts "Осталось бросков: #{@rolls_left-1}"
          print "Бросить ещё? (y/n): "
          ans = gets.chomp.strip.downcase
          if ans == 'y'
            roll_dice
            show_dice
          else
            break
          end
        else
          break
        end
      end
    end

    if @mode == 'ai' && player_idx == 1
      cat, score = best_category(@dice, player_idx)
      if cat
        puts "Компьютер выбрал категорию: #{cat} (+#{score} очков)"
        p[:categories][cat] = score
        p[:score] += score
      else
        puts "Нет доступных категорий. Ход пропущен."
      end
    else
      show_categories(player_idx)
      loop do
        print "Выберите категорию (номер): "
        line = gets.chomp.strip
        exit if line == 'q'
        idx = line.to_i - 1
        if idx >= 0 && idx < CATEGORIES.size
          cat = CATEGORIES[idx]
          if p[:categories][cat]
            puts "Эта категория уже использована."
            next
          end
          score = score_category(@dice, cat)
          p[:categories][cat] = score
          p[:score] += score
          puts "Категория '#{cat}' записана! +#{score} очков"
          break
        else
          puts "Неверный номер."
        end
      end
    end
  end

  def play
    puts colorize("🎲 Добро пожаловать в Яхтзи (покер на костях)!", :bold)
    puts "Правила: бросайте 5 кубиков, выбирайте лучшую комбинацию."
    puts "У вас 3 броска за ход.\n"

    until @game_over
      p = @players[@current]
      if p[:categories].size == CATEGORIES.size
        @current = (@current + 1) % @players.size
        if @players.all? { |pl| pl[:categories].size == CATEGORIES.size }
          @game_over = true
          next
        end
        next
      end
      take_turn(@current)
      @current = (@current + 1) % @players.size
    end

    puts "\n" + colorize("🏆 ИГРА ЗАВЕРШЕНА!", :bold)
    @players.each { |p| puts "#{p[:name]}: #{p[:score]} очков" }
    winner = @players.max_by { |p| p[:score] }
    puts colorize("Победил #{winner[:name]}!", :green)
    @stats['games'] += 1
    @stats['wins'][winner[:name]] ||= 0
    @stats['wins'][winner[:name]] += 1
    save_stats
  end
end

def main
  mode = 'ai'
  names = nil
  show_stats = false
  reset = false
  i = 0
  while i < ARGV.size
    arg = ARGV[i]
    case arg
    when 'ai', 'vs' then mode = arg
    when '-n'
      names = ARGV[i+1].split(',') if i+1 < ARGV.size
      i += 1
    when '-s' then show_stats = true
    when '-r' then reset = true
    when '-h'
      puts "Usage: ruby yahtzee.rb [ai|vs] [-n names] [-s] [-r]"
      return
    end
    i += 1
  end
  if reset
    f = File.join(Dir.home, '.yahtzee_stats.json')
    File.delete(f) if File.exist?(f)
    puts "Статистика сброшена."
    return
  end
  game = Yahtzee.new(mode, names)
  if show_stats
    game.display_stats
    return
  end
  game.play
end

main if __FILE__ == $0

File: yahtzee.py
# yahtzee.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import json
import random
from pathlib import Path

# ANSI-цвета
COLORS = {
    'reset': '\033[0m',
    'red': '\033[91m',
    'green': '\033[92m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'magenta': '\033[95m',
    'cyan': '\033[96m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

DICE_EMOJI = ['⚀', '⚁', '⚂', '⚃', '⚄', '⚅']

class Yahtzee:
    CATEGORIES = [
        "единицы", "двойки", "тройки", "четвёрки", "пятёрки", "шестёрки",
        "три одинаковых", "четыре одинаковых", "фулл-хаус",
        "малый стрит", "большой стрит", "шанс", "яхтзи"
    ]

    def __init__(self, mode='ai', names=None):
        self.mode = mode
        self.names = names or ["Игрок 1", "Игрок 2" if mode == 'vs' else "Компьютер"]
        self.players = []
        for name in self.names:
            self.players.append({'name': name, 'score': 0, 'categories': {}})
        self.current_player = 0
        self.dice = [0] * 5
        self.rolls_left = 3
        self.keep = [False] * 5
        self.game_over = False
        self.stats_file = Path.home() / '.yahtzee_stats.json'
        self.load_stats()

    def load_stats(self):
        if self.stats_file.exists():
            with open(self.stats_file, 'r') as f:
                self.stats = json.load(f)
        else:
            self.stats = {'games': 0, 'wins': {}}

    def save_stats(self):
        with open(self.stats_file, 'w') as f:
            json.dump(self.stats, f, indent=2)

    def roll_dice(self):
        for i in range(5):
            if not self.keep[i]:
                self.dice[i] = random.randint(1, 6)
        self.rolls_left -= 1

    def get_dice_display(self):
        return ' '.join(colorize(f"{DICE_EMOJI[d-1]}{d}", ['red','green','yellow','blue','magenta','cyan'][d-1]) for d in self.dice)

    def show_dice(self):
        print(f"Кости: {self.get_dice_display()}")

    def show_categories(self, player_idx):
        player = self.players[player_idx]
        print("\nДоступные категории:")
        for i, cat in enumerate(self.CATEGORIES):
            if cat not in player['categories']:
                print(f"  {i+1}. {cat}")

    def score_category(self, dice, category):
        counts = [0] * 7
        for d in dice:
            counts[d] += 1
        total = sum(dice)
        if category == "единицы": return counts[1] * 1
        elif category == "двойки": return counts[2] * 2
        elif category == "тройки": return counts[3] * 3
        elif category == "четвёрки": return counts[4] * 4
        elif category == "пятёрки": return counts[5] * 5
        elif category == "шестёрки": return counts[6] * 6
        elif category == "три одинаковых":
            if any(c >= 3 for c in counts): return total
            return 0
        elif category == "четыре одинаковых":
            if any(c >= 4 for c in counts): return total
            return 0
        elif category == "фулл-хаус":
            if (3 in counts and 2 in counts) or (5 in counts):
                return 25
            return 0
        elif category == "малый стрит":
            # Проверяем наличие 4 последовательных
            for start in range(1, 4):
                if all(counts[start+i] > 0 for i in range(4)):
                    return 30
            return 0
        elif category == "большой стрит":
            if all(counts[i] > 0 for i in range(1, 6)):
                return 40
            if all(counts[i] > 0 for i in range(2, 7)):
                return 40
            return 0
        elif category == "шанс":
            return total
        elif category == "яхтзи":
            if any(c == 5 for c in counts):
                return 50
            return 0
        return 0

    def best_category(self, dice):
        best_score = -1
        best_cat = None
        for cat in self.CATEGORIES:
            if cat in self.players[self.current_player]['categories']:
                continue
            score = self.score_category(dice, cat)
            if score > best_score:
                best_score = score
                best_cat = cat
        return best_cat, best_score

    def take_turn(self, player_idx):
        player = self.players[player_idx]
        print(f"\n{'='*40}")
        print(colorize(f"Ход: {player['name']}", 'bold'))

        self.rolls_left = 3
        self.keep = [False] * 5
        self.roll_dice()
        self.show_dice()

        while self.rolls_left > 0:
            if self.mode == 'ai' and player_idx == 1:
                # AI выбирает, какие кости оставить (простая стратегия: оставляет те, что дают лучшую комбинацию)
                # Для простоты AI не перебирает все варианты, а просто оставляет большинство одинаковых.
                counts = [0] * 7
                for d in self.dice:
                    counts[d] += 1
                max_count = max(counts)
                if max_count >= 2:
                    target = counts.index(max_count)
                    self.keep = [d == target for d in self.dice]
                else:
                    self.keep = [False] * 5
                # AI делает второй бросок
                self.roll_dice()
                self.show_dice()
                # AI не делает третий бросок для простоты
                break
            else:
                # Игрок выбирает кости для удержания
                choice = input("Введите номера кубиков для удержания (через пробел, 0 - оставить все): ").strip()
                if choice == 'q':
                    sys.exit(0)
                if choice == '0':
                    self.keep = [True] * 5
                else:
                    self.keep = [False] * 5
                    for idx in choice.split():
                        try:
                            i = int(idx) - 1
                            if 0 <= i < 5:
                                self.keep[i] = True
                        except:
                            pass
                if self.rolls_left > 1:
                    print(f"Осталось бросков: {self.rolls_left-1}")
                    choice = input("Бросить ещё? (y/n): ").strip().lower()
                    if choice == 'y':
                        self.roll_dice()
                        self.show_dice()
                    else:
                        break
                else:
                    break

        # Выбор категории
        if self.mode == 'ai' and player_idx == 1:
            cat, score = self.best_category(self.dice)
            if cat:
                print(f"Компьютер выбрал категорию: {cat} (+{score} очков)")
                player['categories'][cat] = score
                player['score'] += score
            else:
                print("Нет доступных категорий. Ход пропущен.")
        else:
            self.show_categories(player_idx)
            while True:
                choice = input("Выберите категорию (номер): ").strip()
                if choice == 'q':
                    sys.exit(0)
                try:
                    idx = int(choice) - 1
                    if 0 <= idx < len(self.CATEGORIES):
                        cat = self.CATEGORIES[idx]
                        if cat not in player['categories']:
                            score = self.score_category(self.dice, cat)
                            player['categories'][cat] = score
                            player['score'] += score
                            print(f"Категория '{cat}' записана! +{score} очков")
                            break
                        else:
                            print("Эта категория уже использована.")
                    else:
                        print("Неверный номер.")
                except:
                    print("Введите число.")

    def play(self):
        print(colorize("🎲 Добро пожаловать в Яхтзи (покер на костях)!", 'bold'))
        print("Правила: бросайте 5 кубиков, выбирайте лучшую комбинацию.")
        print("У вас 3 броска за ход.\n")

        while not self.game_over:
            player = self.players[self.current_player]
            # Проверяем, заполнены ли все категории
            if len(player['categories']) == len(self.CATEGORIES):
                self.current_player = (self.current_player + 1) % len(self.players)
                # Проверяем, все ли игроки завершили
                if all(len(p['categories']) == len(self.CATEGORIES) for p in self.players):
                    self.game_over = True
                continue
            self.take_turn(self.current_player)
            self.current_player = (self.current_player + 1) % len(self.players)

        # Итоги
        print("\n" + colorize("🏆 ИГРА ЗАВЕРШЕНА!", 'bold'))
        for p in self.players:
            print(f"{p['name']}: {p['score']} очков")
        winner = max(self.players, key=lambda p: p['score'])
        print(colorize(f"Победил {winner['name']}!", 'green'))
        # Обновление статистики
        self.stats['games'] += 1
        if winner['name'] not in self.stats['wins']:
            self.stats['wins'][winner['name']] = 0
        self.stats['wins'][winner['name']] += 1
        self.save_stats()

    def display_stats(self):
        if not self.stats or self.stats.get('games', 0) == 0:
            print(colorize("Статистика пуста.", 'yellow'))
            return
        print(colorize("📊 Статистика:", 'bold'))
        print(f"  Всего игр: {self.stats['games']}")
        for name, wins in self.stats.get('wins', {}).items():
            print(f"  {name}: {wins} побед")

def main():
    mode = 'ai'
    names = None
    show_stats = False
    reset = False
    args = sys.argv[1:]
    i = 0
    while i < len(args):
        arg = args[i]
        if arg == 'ai' or arg == 'vs':
            mode = arg
        elif arg == '-n' and i+1 < len(args):
            names = [n.strip() for n in args[i+1].split(',')]
            i += 1
        elif arg == '-s' or arg == '--stats':
            show_stats = True
        elif arg == '-r' or arg == '--reset':
            reset = True
        elif arg == '-h' or arg == '--help':
            print("Usage: yahtzee.py [ai|vs] [-n names] [-s] [-r]")
            return
        i += 1
    if reset:
        stats_file = Path.home() / '.yahtzee_stats.json'
        if stats_file.exists():
            stats_file.unlink()
        print("Статистика сброшена.")
        return
    game = Yahtzee(mode, names)
    if show_stats:
        game.display_stats()
        return
    game.play()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(colorize("\nИгра прервана.", 'yellow'))
        sys.exit(0)
